lib/EmitVerilog/EmitVerilog.cpp

```diff
@@ -1922,93 +1971,101 @@ void ModuleEmitter::collectNamesEmitDecls(Block &block) {
-    // base name may be unique but the suffixed names may not be.  The right
-    // way to solve this is to change the instances and mems in a new Verilog
-    // dialect to use multiple return values, exposing the independent
-    // Value's.
+      // FIXME(verilog dialect): This can cause name collisions, because the
+      // base name may be unique but the suffixed names may not be.  The right
+      // way to solve this is to change the instances and mems in a new
+      // Verilog dialect to use multiple return values, exposing the
+      // independent Value's.
 
-    // Determine what kind of thing this is, and how much space it needs.
-    maxDeclNameWidth =
-        std::max(getVerilogDeclWord(&op).size(), maxDeclNameWidth);
+      // Determine what kind of thing this is, and how much space it needs.
+      maxDeclNameWidth =
+          std::max(getVerilogDeclWord(&op).size(), maxDeclNameWidth);
 
-    // Flatten the type for processing of each individual element.
-    fieldTypes.clear();
-    flattenBundleTypes(result.getType(), "", false, fieldTypes);
+      // Flatten the type for processing of each individual element.
+      fieldTypes.clear();
+      flattenBundleTypes(result.getType(), "", false, fieldTypes);
 
-    // Handle the reg declaration for a memory specially.
-    if (auto memOp = dyn_cast<MemOp>(&op))
-      if (auto dataType = memOp.getDataTypeOrNull())
-        flattenBundleTypes(dataType, "", false, fieldTypes);
+      // Handle the reg declaration for a memory specially.
+      if (auto memOp = dyn_cast<MemOp>(&op))
+        if (auto dataType = memOp.getDataTypeOrNull())
+          flattenBundleTypes(dataType, "", false, fieldTypes);
 
-    for (const auto &elt : fieldTypes) {
-      int bitWidth = getBitWidthOrSentinel(elt.type);
-      if (bitWidth == -1) {
-        emitError(&op, getName(result))
-            << elt.suffix << " has an unsupported verilog type " << elt.type;
-        result = Value();
-        break;
-      }
+      for (const auto &elt : fieldTypes) {
+        int bitWidth = getBitWidthOrSentinel(elt.type);
+        if (bitWidth == -1) {
+          emitError(&op, getName(result))
+              << elt.suffix << " has an unsupported verilog type " << elt.type;
+          result = OpResult();
+          break;
+        }
 
-      if (bitWidth != 1) { // Width 1 is implicit.
-        // Add 5 to count the width of the "[:0] ".
-        size_t thisWidth = getPrintedIntWidth(bitWidth - 1) + 5;
-        maxTypeWidth = std::max(thisWidth, maxTypeWidth);
+        if (bitWidth != 1) { // Width 1 is implicit.
+          // Add 5 to count the width of the "[:0] ".
+          size_t thisWidth = getPrintedIntWidth(bitWidth - 1) + 5;
+          maxTypeWidth = std::max(thisWidth, maxTypeWidth);
+        }
       }
-    }
 
-    // Emit this declaration.
-    if (result)
-      declsToEmit.push_back(&op);
+      // Emit this declaration.
+      if (result)
+        declsToEmit.push_back(result);
```

Comment:
I would recommend keeping `declsToEmit` as a per-op thing, not a per-result thing.

Reply:
I went this way since what this is really doing is declaring wires or regs. But it does munge in some FIRRTL op specifics... This entire method is difficult to understand since it does way too much IMO. It's also going to need some refactoring when it comes time to output actual SV types. RTLInstanceOp only needs it to declare wires... Should I just break that out and mark this function for refactoring?

This actually ties in to a discussion I want to continue tomorrow: whether to support SystemVerilog output via an Interface or lower to the SV dialect and use that to do a very simple conversion. I'll add some text to the meeting notes on my thoughts.

Reply:
Yeah this logic will get a lot cleaner when direct support for firrtl is removed from it.  Feel free to clone this logic into a separate path and send all the firrtl stuff down this path if it makes it cleaner.